--- a/lua/ubus.c	2015-06-16 09:43:49.000000000 +0800
+++ b/lua/ubus.c	2015-09-09 14:33:13.000000000 +0800
@@ -104,7 +104,7 @@
 		break;
 
 	case BLOBMSG_TYPE_STRING:
-		lua_pushstring(L, data);
+		lua_pushlstring( L, data, len-1 );
 		break;
 
 	case BLOBMSG_TYPE_ARRAY:
@@ -181,6 +181,14 @@
 		break;
 
 	case LUA_TSTRING:
+	    {
+	        size_t  tlen;
+	        const char * ptr;
+	        ptr = lua_tolstring( L, -1, &tlen );
+	        blobmsg_add_field( b, BLOBMSG_TYPE_STRING, key, ptr, tlen + 1 );
+	    }
+	    break;
+	    
 	case LUA_TUSERDATA:
 	case LUA_TLIGHTUSERDATA:
 		blobmsg_add_string(b, key, lua_tostring(L, -1));
