--- a/drivers/tty/pty.c
+++ b/drivers/tty/pty.c
@@ -558,6 +558,7 @@ static int pty_unix98_ioctl(struct tty_s
 	case TIOCGPKT: /* Get PT packet mode */
 		return pty_get_pktmode(tty, (int __user *)arg);
 	case TIOCGPTN: /* Get PT Number */
+                clear_bit(TTY_OTHER_CLOSED, &tty->flags);
 		return put_user(tty->index, (unsigned int __user *)arg);
 	case TIOCSIG:    /* Send signal to other side of pty */
 		return pty_signal(tty, (int) arg);
